File: backend/internal/domain/slot.go
package domain

import "context"

type Slot struct {
	ID              int64    `db:"id" json:"id"`
	DateTime        string   `db:"date_time" json:"dateTime"`
	EndTime         string   `db:"end_time" json:"endTime"`
	ProgramID       int64    `db:"program_id" json:"-"`
	MasterID        int64    `db:"master_id" json:"-"`
	TotalSpots      int      `db:"total_spots" json:"totalSpots"`
	AvailableSpots  int      `db:"available_spots" json:"availableSpots"`
	RentalAvailable bool     `db:"rental_available" json:"rentalAvailable"`
	RentalPrice     float64  `db:"rental_price" json:"rentalPrice"`
	Program         *Program `json:"program,omitempty"`
	Master          *Master  `json:"master,omitempty"`
}

type SlotFilter struct {
	DateFrom  string
	DateTo    string
	MasterID  *int64
	ProgramID *int64
}

type SlotRepository interface {
	QuerySlots(ctx context.Context, filter SlotFilter) ([]Slot, error)
	GetSlotByID(ctx context.Context, id int64) (Slot, error)
	DecrementSpots(ctx context.Context, slotID int64) error
	IncrementSpots(ctx context.Context, slotID int64) error
}

File: mobile/lib/core/models/slot.dart
import 'program.dart';
import 'master.dart';

class Slot {
  final int id;
  final String dateTime;
  final String endTime;
  final Program program;
  final Master master;
  final int totalSpots;
  final int availableSpots;
  final bool rentalAvailable;
  final double rentalPrice;

  const Slot({
    required this.id,
    required this.dateTime,
    required this.endTime,
    required this.program,
    required this.master,
    required this.totalSpots,
    required this.availableSpots,
    required this.rentalAvailable,
    required this.rentalPrice,
  });

  bool get isAvailable => availableSpots > 0;

  factory Slot.fromJson(Map<String, dynamic> json) => Slot(
    id: json['id'] as int,
    dateTime: json['dateTime'] as String,
    endTime: json['endTime'] as String,
    program: Program.fromJson(json['program'] as Map<String, dynamic>),
    master: Master.fromJson(json['master'] as Map<String, dynamic>),
    totalSpots: json['totalSpots'] as int,
    availableSpots: json['availableSpots'] as int,
    rentalAvailable: json['rentalAvailable'] as bool,
    rentalPrice: (json['rentalPrice'] as num).toDouble(),
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'dateTime': dateTime,
    'endTime': endTime,
    'program': program.toJson(),
    'master': master.toJson(),
    'totalSpots': totalSpots,
    'availableSpots': availableSpots,
    'rentalAvailable': rentalAvailable,
    'rentalPrice': rentalPrice,
  };
}
